### ISSUE TITLE
(inactive) DIscBIO

### ISSUE THREAD

**Issue by wleoncio** [Bioconductor/Contributions#1376]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/ocbe-uio/DIscBIO

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by mtmorgan** [comment]
This package should almost certainly implement an interface to [SingleCellExperiment][1] and use other [standard Bioconductor representations][2] for robustness and interoperability; see for instance this [recent comment][3].

[1]: https://bioconductor.org/packages/SingleCellExperiment
[2]: https://bioconductor.org/developers/how-to/commonMethodsAndClasses/
[3]: https://github.com/Bioconductor/Contributions/issues/1355#issuecomment-580977106

**Comment by wleoncio** [comment]
> On one or more platforms, the build results were: "skipped, ERROR".

Is it possible to trigger a new build? The error seems to be caused by a failed attempt to access the string-db.org API. As further example that the package should be building fine, here is the TravisCI output from a build three days ago.

![bilde](https://user-images.githubusercontent.com/8234768/75665757-f7b4bb80-5c74-11ea-9ff5-afec0c6e88e8.png)

**Comment by lshep** [comment]
A new build is triggered with a version bump in the DESCRIPTION if your [webhook is set up](https://github.com/Bioconductor/Contributions/blob/master/CONTRIBUTING.md#adding-a-web-hook).  And please see @mtmorgan comment about changing the interfaces utiltized

**Comment by wleoncio** [comment]
> A new build is triggered with a version bump in the DESCRIPTION if your [webhook is set up](https://github.com/Bioconductor/Contributions/blob/master/CONTRIBUTING.md#adding-a-web-hook).

Got it, I'll take care of that.

>  And please see @mtmorgan comment about changing the interfaces utiltized

I defer a reply to the package's principal contributor, @systemsbiologist, who can better explain the choice to use a new class instead of the ones available at [SingleCellExperiment](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html).

**Comment by SystemsBiologist** [comment]
Dear @lshep and @mtmorgan,

Thank you very much for your feedback. We appreciate the points you raised about interoperability, and expectation that all Bioconductor packages should work together. We took on your suggestion, and started developing an interface that allows SingleCellExperiment-class objects to be analyzed using our package. We will consider using other standard Bioconductor representations wherever possible/appropriate. We are planning to resubmit an updated version of our package within 1-2 weeks. Thanks .

**Comment by hpages** [comment]
Thanks @SystemsBiologist and @wleoncio for the update.

H.

**Comment by hpages** [comment]
Hi @wleoncio and @SystemsBiologist ,

Do you have any update on this?

Thanks,
H.

**Comment by wleoncio** [comment]
Hi @hpages, thanks for checking in! We have finished implementing all changes requested and are currently adding some unit tests and making sure everything is working as expected. We might go a few days over the previous 1-2 weeks prediction, but we're nearly ready to push a new patch to master.

**Comment by wleoncio** [comment]
The warnings are quite minor ("update R from 3.6 to 4.0", "add blank line to end of the file") and I am ready to push the fixes to master. But I am a bit stumped about the Windows error:

```
 ** byte-compile and prepare package for lazy loading
 Error: .onLoad failed in loadNamespace() for 'rJava', details:
  call: library.dynam("rJava", pkgname, libname)
  error: DLL 'rJava' not found: maybe not installed for this architecture?
```

Should I worry about this or is this likely to be a fluke?

**Comment by lshep** [comment]
I reinstalled rJava yesterday afternoon on the builders so you should hopefully not have this issue anymore

**Comment by wleoncio** [comment]
Finally! Now I can ~go out~ stay indoors for Easter. Happy holidays to whoever reads this! :)

**Comment by SystemsBiologist** [comment]
Happy Easter!

**Comment by hpages** [comment]
Hi @SystemsBiologist, @wleoncio,

How is the refactoring of the package going? As of today, I see that the pipeline implemented in the package is still entirely based on DIscBIO objects, which are incompatible with SingleCellExperiment objects. FWIW Martin Morgan's suggestion (@mtmorgan) was that you use SingleCellExperiment objects instead for interoperability with all other single-cell analysis packages in Bioconductor.

Please don't hesitate to ask on the bioc-devel mailing list if you have questions or concerns about this. Thanks!

H.

**Comment by wleoncio** [comment]
Dear @hpages,

Interoperability with the SCE objects is done via the `as.DISCBIO` function, whose documentation contains the following example:

```R
g1_sce <- SingleCellExperiment::SingleCellExperiment(
    list(counts=as.matrix(valuesG1msReduced)) # internal dataset
)
g1_disc <- as.DISCBIO(g1_sce)
class(g1_disc) #returns "DISCBIO"
```

That function basically imports an SCE object and transforms the slots accordingly for usage on DIscBIO. Is that a reasonable solution? I'm afraid refactoring all functions by, say, adding methods for SCE objects on each of them would be quite the undertaking, which could set us back a few months.

**Comment by hpages** [comment]
In practice this approach won't provide much interoperability with other single-cell analysis packages in Bioconductor, for a number of reasons. For example `as.DISCBIO()` is largely untested and will fail for most SingleCellExperiment objects (except for the trivial one you use in your example). Also at any point in your pipeline the user should be able to turn their DISCBIO object back into a SingleCellExperiment object. Otherwise, they'll be stuck in the DIscBIO pipeline with their DISCBIO object. _Interoperability_ with other tools means that at any point in their analysis, the user can replace a tool with another if they want.

The deeper issue is that the DISCBIO container is re-inventing the wheel. On the other hand Bioconductor already has a well-tested/well-supported container for representing single-cell data. So, yes, our suggestion is that there is no need for the DISCBIO container and that you base your pipeline on SingleCellExperiment objects.

Thanks,
H.

**Comment by wleoncio** [comment]
Thank you for the clarification, @hpages. I'll discuss the issue with @SystemsBiologist, we'll act on it ASAP.

**Comment by SystemsBiologist** [comment]
Dear @hpages,

Thanks for your reply. We understand you are not happy with the updated package. Please, rest assured that we are not trying to re-invent the wheel, and we agree that the singleCellExperiment-class should be the reference class in any single cell analysis pipeline. We understand your points but it seems to us you changed your mind about an interface being sufficient in our case.

We are really willing to put an extra effort and revise again our package to make it suitable for the bioconductor ecosystem. However, we need to make sure that our work will be aligned to the requirements for acceptance in bioconductor. This is our plan. We can definitely use a singeCellExperiment object as the only input of our pipeline, yet we need some custom/extra slots for storing intermediates and results of our analysis. We can rewrite the DISCBIO class definition so that the DISCBIO-class can>
 1) extend the singleCellExperiment-class, or  
2) just include a slot for a singleCellExperiment object. Would either strategy be compatible with bioconductor requirements? Please, let us know.

Best regards

**Comment by hpages** [comment]
> but it seems to us you changed your mind about an interface being sufficient in our case

There is no change of mind. Maybe we were not clear enough initially about what we expected? Sorry if that was the case. Just to clarify, rest assured that Bioconductor commitment to standard representations and interoperability has always been **very strong**.

We sometimes refer to the 2 approaches you describe as using _inheritance_ vs _composition_. Most of the times developers try to go for the 1st, and only if they run into problems, they switch to the 2nd. The problem one typically faces with the _inheritance_ approach is that they have to choose which SummarizedExperiment subclass to derive from. For example in your case the SingleCellExperiment class. If in the future you decide that you want to support other SingleCellExperiment derivatives, then you'll have to introduce a new subclass for each of them. This would quickly become cumbersome and hard to maintain. Using _composition_ avoids that problem e.g. if you include a slot that expects a SingleCellExperiment object then it would accept a SingleCellExperiment object or derivative. However, the big drawback of the _composition_ approach is that your objects don't inherit the SingleCellExperiment API. But since it's easy to get the SingleCellExperiment object out of the bigger object, this should not be too bad in practice.

A 3rd approach is to not introduce a new class and to just use a plain SingleCellExperiment object. And then to store the intermediate results that don't fit in the `assays` or `rowData` or `colData` components of the object in its `metadata` component.

Hope this helps,
H.

**Comment by SystemsBiologist** [comment]
Dear @hpages,

Thanks for your clarification. We are wondering if we go with the second option (rewriting the DISCBIO class definition to include a slot for a singleCellExperiment object) would that be compatible with Bioconductor requirements? Or we should go with the third option (using a plain SingleCellExperiment object. and storing the intermediate results in the metadata component)?

Looking forward to hearing from you

**Comment by hpages** [comment]
Using a plain SingleCellExperiment object, if possible, is the preferred way. If your intermediate results have the rectangular shape of the SingleCellExperiment object they could go in an assay. If they are _parallel_ to its 1st dimension, they should go in the rowData. If they are _parallel_ to its 2nd dimension, they should go in the colData. If they have a rectangular shape with the same number of columns as the object but not necessarily the same number of rows, they could go in the altExps list. If they don't fit in any of these components, they can always go in the metadata list. But this should typically be the last resort.

Several pipelines in Bioconductor start with a SingleCellExperiment object (or other SummarizedExperiment derivative) that contains only 1 assay with the raw counts, and then stuff the object with intermediate results as we move thru the pipeline. I suggest that you take a look at what other packages do to get a feeling of whether or not that would work for you. If for whatever reason it doesn't, then I guess it would be ok to go for the second option (_composition_), even though that is not how other pipelines typically handle this.

I'm by no mean a SingleCellExperiment pipeline expert so please don't hesitate to seek advice on the bioc-devel mailing list or the Bioconductor Community Slack (#SingleCellExperiment channel, see Community Slack sign-up on [our website](https://bioconductor.org/)) where the real experts should be able to give valuable advice.

Best,
H.

**Comment by wleoncio** [comment]
The error seems like a connection issue ("GnuTLS recv error (-110): The TLS connection was non-properly terminated."). Is it possible to trigger another build without committing on our master branch?

**Comment by lshep** [comment]
Normally only commits to the master branch will trigger builds.  We do have a way to do this manually.  I will trigger a build manually this once but please don't rely on this often.  (your report should post shortly)

**Comment by wleoncio** [comment]
@lshep, thank you for making an exception. Too bad it triggered the same error. We're going to take another look in our end, we can try to add some code to circumvent this issue on the package side.

**Comment by SystemsBiologist** [comment]
Dear @hpages,

We submitted an updated version of our DISCBIO R package. After your feedback, we assessed the feasibility of using the singleCellExperiment-class as the sole class to store all data and intermediates from our pipeline. However, we concluded that this was not the best solution here. Our pipeline is mainly designed to process single cell experiments that include spike-in samples. These samples are used for normalization and internal control, and need to be isolated from the rest of the data at the very beginning of our pipeline. Since this operation generates matrices with sizes that do not match (nor match with the input expression counts matrix), it would be inefficient to store the DISCBIO intermediate data in a single singleCellExperiment object (it is feasible, but the resulting structure would be complex and counter-intuitive, and may lead to significant data duplication depending on the number of spike-ins).

On the contrary, we considered that a more efficient approach was to rely on composition (that you also recommended as an alternative scenario). The DISCBIO pipeline now starts from a singleCellExperiment object that provides the input data. Specifically, a singleCellExperiment object is stored as one of the slots of our DISCBIO class. Expression counts are retrieved from the singleCellExperiment using singleCellExperiment-specific methods, and then spike-in samples are extracted (during the second step) to generate the reduced-size matrices that are used in the following steps of our pipeline.

Since the updated version of DISCBIO relies on composition, importing a singleCellExperiment into a DISCBIO object (or exporting from) is trivial, thus facilitating the processing of the same data via alternative Bioconductor pipelines. We think that this solution satisfies Bioconductor’s guidelines aimed at enhancing code reuse and interoperability.

We thank you in advance for your time and for considering DISCBIO for inclusion in Bioconductor.

Looking forward to hearing from you

**Comment by hpages** [comment]
Hi,

Have you considered storing the spike-in data in the altExps component of the SingleCellExperiment object? This is the typical use case that motivated the introduction of the altExps feature by the SingleCellExperiment developers. As mentioned earlier, the experts on the  #SingleCellExperiment channel of the community-bioc Slack will happily provide advice on these matters.

Thanks

**Comment by wleoncio** [comment]
Dear @hpages,

We have discussed the suggested implementation; unless we misunderstood the final goal of the review board, we are afraid this would trigger an overhaul of the package to accomodate a SingleCellExperiment object into the whole DIscBIO pipeline. Our package was envisioned as a standalone pipeline and we are definitely willing to integrate the initial input and final output with other classes from the Bioconductor universe, but would rather keep the intermediate results (including spike-in calculations) within the DISCBIO-class object. I hope this is not an impediment to the acceptance of the package, but if it is we totally understand and are prepared to retract the submission.

**Comment by SystemsBiologist** [comment]
Dear @hpages,

We hope you are doing great!
We are just wondering if you have had the chance to take a decision regarding our package, which we really hope it gets accepted and be part of Bioconductor.

Looking forward to hearing from you

**Comment by hpages** [comment]
Hi,

So I took a look at your DISCBIO object (21 slots!) and here is what I see:

**Initial DISCBIO object**

We enter the pipeline by feeding the `DISCBIO()` constructor with an M x N data frame containing single cell data with N samples and M features (genes + ERCC spike-ins).

This populates slots `SingleCellExperiment` (SingleCellExperiment object, a.k.a. SCE) and `expdataAll` (data frame), each with a full copy of the input data, It also populates slots `expdata`, `ndata`, and `fdata` (all of them data frames), each with a copy of the input data from which the ERCC spike-ins have been removed.

The `altExp` feature of the SCE should be used to store the spike-in data separately from the main data. Then the other 5 slots become redundant and should be removed.

**NoiseFiltering step**

This steps populates slot `noiseF` with the names of the genes that passed the filtering.

This should be stored in the SCE object as a column (typically logical) of its `rowData()` component. Then the `noiseF` slot becomes redundant and should be removed.

**Normalizedata step**

This steps populates slots `ndata` and `fdata` with the normalized data (the 2 slots are populated with identical data), and slot `filterpar` with some parameters (which, according to the man page, are filtering parameters but the slot only got populated now and not by the NoiseFiltering step).

The normalized data should be stored in the SCE object as a new assay, and the list of parameters should go in its `metadata()` component. Then slots `ndata`, `fdata`, and `filterpar` become redundant and should be removed.

**FinalPreprocessing step**

This is conceptually a subsetting operation where only a subset of the initial genes is kept. The way this is implemented in the DISCO object is by populating slot `FinalGeneList` with a copy of the gene list that is already in `noiseF`, and by populating slot `fdata` with `ndata[FinalGeneList, ]`.

With the SCE object, this is just:

    SCE <- SCE[rowData(SCE)$noiseF, ]

granted that `rowData(SCE)$noiseF` is the logical column containing the result of the NoiseFiltering step.

**Clustexp step**

This steps populates slots `distances`, `cpart`, `clusterpar`, `kmeans`, and `fcol`. `distances` is a square matrix representing the pairwise distances between the samples. Storing an object that is going to grow quadratically with the number of samples will certainly cause problems with most real world single cell datasets where it's not uncommon to see tens of thousands of samples. Not a problem of course in the examples provided in the vignette and man pages where the number of samples is 30.

Anyway this matrix should be stored in the SCE object as a column of its `colData()` component. Same thing for the `cpart` slot which is an integer vector of length N. The lists in `clusterpar` and `kmeans` should go in its `metadata()` component. Same thing for the cluster color scheme in the `fcol` slot.

Then these 5 slots are not longer needed and should be removed.

**comptSNE step**

This steps populates the `tsne` slot with an N x 2 data frame.

No need for this slot either. This should go in the SCE object as a column of its `colData()` component.

etc...

The bottom line is that the DISCBIO container is unfortunately a clumsy attempt at reinventing a container that already exists in Bioconductor. I can understand that modifying the DIscBIO pipeline to accommodate a SingleCellExperiment object represents a major overhaul but, as mentioned previously, contributions to the project are required to adopt standard representations and provide interoperability with the other tools in the ecosystem.

Let us know what you want to do.

Regards,
H.

**Comment by wleoncio** [comment]
Dear Reviewer,

Thank you for taking the time to provide such thorough analysis and comments that would surely improve our package. 

Unfortunately, we cannot currently implement such changes without critically compromising our available resources. We will, however, take all your notes into consideration should the opportunity to publish a future version of DIscBIO on Bioconductor arise.

All the best,
Waldir (on behalf of all package authors)
